File: apps/web/src/lib/api/schemas/custom_domains.ts
import { z } from "zod";

export const customDomainPhaseSchema = z.enum([
  "awaiting_verification",
  "validating",
  "ready",
  "action_required",
]);
export type CustomDomainPhase = z.infer<typeof customDomainPhaseSchema>;

export const customDomainErrorCodeSchema = z.enum([
  "verification_timed_out",
  "certificate_failed",
  "hostname_rejected",
  "upstream_unavailable",
]);
export type CustomDomainErrorCode = z.infer<typeof customDomainErrorCodeSchema>;

export const customDomainDnsModeSchema = z.enum(["cname", "delegation"]);
export type CustomDomainDnsMode = z.infer<typeof customDomainDnsModeSchema>;

export const dnsRecordSchema = z.object({
  type: z.string(),
  name: z.string(),
  value: z.string(),
});
export type DnsRecord = z.infer<typeof dnsRecordSchema>;

export const customDomainSchema = z.object({
  id: z.string(),
  hostname: z.string(),
  dns_mode: customDomainDnsModeSchema,
  phase: customDomainPhaseSchema,
  required_records: z.array(dnsRecordSchema).default([]),
  error_code: customDomainErrorCodeSchema.nullable().default(null),
  error_message: z.string().nullable().default(null),
  verified_at: z.string().datetime().nullable().default(null),
  last_checked_at: z.string().datetime().nullable().default(null),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});
export type CustomDomain = z.infer<typeof customDomainSchema>;

export const customDomainListSchema = z.object({
  data: z.array(customDomainSchema).default([]),
  next: z.string().default(""),
});
export type CustomDomainList = z.infer<typeof customDomainListSchema>;

File: packages/shared/src/shared/http/custom_domains.py
from __future__ import annotations

from datetime import datetime

from pydantic import Field

from shared.custom_domains import (
    MAX_HOSTNAME_LENGTH,
    CustomDomain,
    CustomDomainDnsMode,
    CustomDomainErrorCode,
    CustomDomainPhase,
    DnsRecord,
)
from shared.http.base import HttpModel


class CustomDomainRegisterRequest(HttpModel):
    domain: str = Field(min_length=3, max_length=MAX_HOSTNAME_LENGTH)
    dns_mode: CustomDomainDnsMode


class CustomDomainResponse(HttpModel):
    id: str
    hostname: str
    dns_mode: CustomDomainDnsMode
    phase: CustomDomainPhase

    required_records: tuple[DnsRecord, ...] = ()
    """The exact DNS records Pangolin is waiting for."""

    error_code: CustomDomainErrorCode | None = None
    error_message: str | None = Field(default=None, max_length=512)
    verified_at: datetime | None = None
    last_checked_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class CustomDomainListResponse(HttpModel):
    data: list[CustomDomainResponse] = Field(default_factory=list)
    next: str = ""


def custom_domain_response(domain: CustomDomain) -> CustomDomainResponse:
    return CustomDomainResponse(
        id=domain.id,
        hostname=domain.hostname,
        dns_mode=domain.dns_mode,
        phase=domain.phase,
        required_records=domain.required_records,
        error_code=domain.error_code,
        error_message=domain.error_message,
        verified_at=domain.verified_at,
        last_checked_at=domain.last_checked_at,
        created_at=domain.created_at,
        updated_at=domain.updated_at,
    )


__all__ = [
    "CustomDomainListResponse",
    "CustomDomainRegisterRequest",
    "CustomDomainResponse",
    "custom_domain_response",
]
